nestkernel/conn_builder.cpp

```diff
@@ -406,111 +421,126 @@ nest::ConnBuilder::single_connect_( index snode_id, Node& target, thread target_
-    {
-      kernel().connection_manager.connect( snode_id,
-        &target,
-        target_thread,
-        synapse_model_id_,
-        dummy_param_dicts_[ target_thread ],
-        delay_->value_double( target_thread, rng, snode_id, &target ) );
-    }
-    else if ( default_delay_ )
-    {
-      kernel().connection_manager.connect( snode_id,
-        &target,
-        target_thread,
-        synapse_model_id_,
-        dummy_param_dicts_[ target_thread ],
-        numerics::nan,
-        weight_->value_double( target_thread, rng, snode_id, &target ) );
+      if ( default_weight_and_delay_[ indx ] )
+      {
+        kernel().connection_manager.connect(
+          snode_id, &target, target_thread, synapse_model_id_[ indx ], dummy_param_dicts_[ target_thread ] );
+      }
+      else if ( default_weight_[ indx ] )
+      {
+        kernel().connection_manager.connect( snode_id,
+          &target,
+          target_thread,
+          synapse_model_id_[ indx ],
+          dummy_param_dicts_[ target_thread ],
+          delay_[ indx ]->value_double( target_thread, rng, snode_id, &target ) );
+      }
+      else if ( default_delay_[ indx ] )
+      {
+        kernel().connection_manager.connect( snode_id,
+          &target,
+          target_thread,
+          synapse_model_id_[ indx ],
+          dummy_param_dicts_[ target_thread ],
+          numerics::nan,
+          weight_[ indx ]->value_double( target_thread, rng, snode_id, &target ) );
+      }
+      else
+      {
+        double delay = delay_[ indx ]->value_double( target_thread, rng, snode_id, &target );
+        double weight = weight_[ indx ]->value_double( target_thread, rng, snode_id, &target );
+        kernel().connection_manager.connect( snode_id,
+          &target,
+          target_thread,
+          synapse_model_id_[ indx ],
+          dummy_param_dicts_[ target_thread ],
+          delay,
+          weight );
+      }
     }
     else
     {
-      double delay = delay_->value_double( target_thread, rng, snode_id, &target );
-      double weight = weight_->value_double( target_thread, rng, snode_id, &target );
-      kernel().connection_manager.connect(
-        snode_id, &target, target_thread, synapse_model_id_, dummy_param_dicts_[ target_thread ], delay, weight );
-    }
-  }
-  else
-  {
-    assert( kernel().vp_manager.get_num_threads() == static_cast< thread >( param_dicts_.size() ) );
+      assert( kernel().vp_manager.get_num_threads() == static_cast< thread >( param_dicts_[ indx ].size() ) );
 
-    ConnParameterMap::const_iterator it = synapse_params_.begin();
-    for ( ; it != synapse_params_.end(); ++it )
-    {
-      if ( it->first == names::receptor_type or it->first == names::music_channel or it->first == names::synapse_label )
+      for ( auto synapse_parameter : synapse_params_[ indx ] )
       {
-        try
+        if ( synapse_parameter.first == names::receptor_type or synapse_parameter.first == names::music_channel
```

Comment:
This code here looks overly complicated and should long have been refactored (I wrote it once upon a time, I believe ...). We can do this *much* simpler: We add to `ConnParameter` methods `provides_int` and `provides_double`, both returning `false` for the base class and `true` where suitable. For random and parameter-wrapper, both return true. Then here, we only need to do
```
if ( synapse_parameter.second->provides_double() )
{  // double is more frequent case, test first
  DoubleDatum* ...
}
else
{
  IntDatum* ...
}
```
We can skip the entire try-catch construction. Nothing can go wrong here if we set up parameters properly in the constructor.